cypress-io/cypress
# feat: add cohorts for A/B testing in Specs List Banners

- Closes https://github.com/cypress-io/cypress/issues/23083
- Closes https://github.com/cypress-io/cypress/issues/23118

### User facing changelog

n/a

### Additional details


In order to be able to A/B test different concepts in the app, an ability to assign a user to a particular cohort is needed.  A cohort can be assigned per feature or functionality that needs to be A/B tested.  

In this first use case, A/B testing is limited to the ACI Smart Prompt banners being shown on the Specs list page of the app.  See the details in https://github.com/cypress-io/cypress/issues/23118 that describes the specific changes.

The cohort functionality is encapsulated in a composable called `useCohorts` in the `frontend-shared` package.  

### Steps to test


#### Prerequisites

The cohort functionality is used within the Smart Prompt banners used on the Specs list page of the app.  In order to fully test the different banners, the project used for testing will need to be in a certain state.

- The project should have been opened more than 4 days ago.  This can be confirmed by looking at the `state.json` file under the "App Data" directory and then navigating to "projects" and then the directory with the project name and hash.  The `firstOpened` value should be a timestamp older than 4 days ago
- Remove the `banners` entry from the same file. This will reset the apps tracking of when banners were viewed or dismissed.
- Make sure the project is not connected to a dashboard project id by removed the "lastProjectId" from the same file and the `cypress.config.js` file for the project as appropriate

#### Testing cohorts

* After first starting the app with the launchpad loaded, select the "Developer tools > View App Data"  menu item
* Within the directory that opens, open the `cache` file in a text editor
* Confirm that there is not a "COHORTS" entry in the json data
* Make sure you are logged out from the dashboard

* Login Banner
	* Open the project in either e2e or component testing mode
	* Confirm that the Login Banner shows
	* Reopen the `cache` file from above and confirm there is now a COHORTS entry with a key of `aci_082022_login` that matches the bannerId in `LoginBanner.vue`.  The key should have an object with the name matching the bannerId and a cohort value of either 'A' or 'B'.  Confirm the body copy of the Login Banner matches the appropriate copy for either A or B shown in the document [here](https://docs.google.com/document/d/1PFN79d3deDyqDVFmVIBe4cVMzWiEll0J6XCpesBZrSg/edit#)
	* Navigate away from the Specs list page by going to the Runs page and then back to the Specs list.  The same copy should be showing in the Login Banner from before and the "COHORTS" value in the `cache` file should not have changed
	* Close the browser for the app either manually or from the launch pad.
	* Edit the `cache` file to remove the key for `aci_082022_login`
	* Open the app again in either mode and confirm that the `cache` file has a new entry for `aci_082022_login` and that the appropriate copy is showing
	* If the algorithm selected the same cohort as before, you can test the other option in two ways:
	  * Follow the steps above for a couple of more times to edit the `cache` file and remove the same entry until the algorithm chooses the other cohort
	  * Edit the `cache` file to manually edit the cohort to the other value (`A` or `B`)
*  Remaining banners
  * Test the other Smart Prompt banners as described here: https://github.com/cypress-io/cypress/pull/23256
  * For each banner, confirm the same "cohort" behavior and compare the results with the expected A/B testing values [here](https://docs.google.com/document/d/1PFN79d3deDyqDVFmVIBe4cVMzWiEll0J6XCpesBZrSg/edit#)
  * The keys in the "COHORTS" section should match the bannerIds for each banner
    * `aci_082022_createOrganization`
    * `aci_082022_connectProject`
    * `aci_082022_record`
* UTM Parameters
  * The calculated cohort values should also be passed with the resulting call to actions from each banner.  Run through the tests for this [PR](https://github.com/cypress-io/cypress/pull/23691) to confirm the cohort values are passed in the external links for the Login and Create Organization and in the mutation for creating a project.  The Record Runs banner does not track the cohort when running the command

### How has the user experience changed?

Once fully implemented, the only user experience change will be that different copy may be used in the banners depending on which cohort the user is assigned to.

### PR Tasks


- [n/a] Has a PR for user-facing changes been opened in [`cypress-documentation`](https://github.com/cypress-io/cypress-documentation)? 
- [n/a] Have API changes been updated in the [`type definitions`](https://github.com/cypress-io/cypress/blob/develop/cli/types/cypress.d.ts)?
